### PR TITLE
Improve segment processing

### PR DESCRIPTION
**Pull request by drinkcoffee**
Changed parameter names to make naming convention.
Removed old comments.
Added revert reason to require.
Checked call path to ensure segment events call path matched the call path of the segment being processed.

The happy path for these changes are tested via the Trade-Finance example.

### CODE REVIEW

File: application/atomic-appcontracts/erc20/src/main/java/net/consensys/gpact/appcontracts/atomic/erc20/soliditywrappers/GpactERC20Bridge.java [hunk 1]
```diff
@@ -40,7 +40,7 @@
...
...653a20426c6f636b636861696e206e6f74207375706040820152651c1bdc9d195960d21b606082015260800190565b6020808252603c908201527f4552433230204272696467653a20546f6b656e206e6f74207472616e7366657260408201527f61626c6520746f2072657175657374656420626c6f636b636861696e00000000606082015260800190565b6001600160a01b039384168152919092166020820152604081019190915260600190565b8381526000602060018060a01b0385168184015260606040840152835180606085015260005b81811015611a8e57858101830151858201608001528201611a72565b81811115611aa0576000608083870101525b50601f01601f19169290920160800195945050505050565b60208082526024908201527f4552433230204272696467653a204d7573742068617665204d415050494e4720604082015263726f6c6560e01b606082015260800190565b60208082526022908201527f4552433230204272696467653a20746f6b656e206e6f7420636f6e6669677572604082015261195960f21b606082015260800190565b60208082526023908201527f4552433230204272696467653a204d75737420686176652050415553455220726040820152626f6c6560e81b606082015260800190565b600060208284031215611b9357600080fd5b8151611730816118a956fe0121836bd496e565c22481bc40c30c41db1fecf07c7376491a377f8056cfccd5a2646970667358221220ace52e9ac48b5a263fa0fa994d382dae158e0fe5e3528024b13bc2ba51d0c49464736f6c63430008090033";
```
Comment (ermyas) [inline application/atomic-appcontracts/erc20/src/main/java/net/consensys/gpact/appcontracts/atomic/erc20/soliditywrappers/GpactERC20Bridge.java:43]: The wrappers appear not to be in sync with the solidity contract changes. Can you build locally and commit the latest wrappers pls?

File: functioncall/gpact/src/main/solidity/CrosschainControl.sol [hunk 2]
```diff
@@ -716,45 +714,48 @@ contract CrosschainControl is
...
             );
 
-            // Segments with offset zero of Root calls are the only ones that can call segments.
-            require(callPath[callPath.length - 1] == 0);
-            //            require(lenOfCallPath == callPath.length || lenOfCallPath == callPath.length+1, "Bad call path length for segment");
+            // The caller must be a segment that is calling other segments or the
+            // root call (which also is calling segments). These both have the call path
+            // for the caller will have 0 as the offset in the call path array for the
+            // previous level in the call path.
+            require(_callPath[_callPath.length - 1] == 0, "Invalid caller");
```
Comment (ermyas) [inline functioncall/gpact/src/main/solidity/CrosschainControl.sol:730]: I might be missing something but is there a reason this `require` is in the for loop? `_callPath` refers to the function arg and does not seem to change during the loop iterations?

**Review by ermyas — APPROVED** [review]
LGTM... minor comments added

**Review by wcgcyx — APPROVED** [review]

File: functioncall/gpact/src/main/solidity/CrosschainControl.sol [hunk 3]
```diff
@@ -716,45 +714,48 @@ contract CrosschainControl is
...
                 "Bad call path length for segment"
             );
-            // TODO walk through call path to ensure the call path from the segment event is correct
-            //            if (lenOfCallPath == callPath.length+1) {
-            //                uint256 callPathFinalValue = BytesUtil.bytesToUint256(segmentEvent, locationOfCallPath + 0x20 * (lenOfCallPath-1));
-            //                require(callPathFinalValue == 0, "Call path optional second element not zero");
-            //            }
+            if (segCallPath.length == _callPath.length + 1) {
+                require(
+                    segCallPath[segCallPath.length - 1] == 0,
```
Comment (ricardolyn) [inline functioncall/gpact/src/main/solidity/CrosschainControl.sol:743]: same question here

File: functioncall/gpact/src/main/solidity/CrosschainControl.sol [hunk 4]
```diff
@@ -716,45 +714,48 @@ contract CrosschainControl is
...
+                    "Final call path element not zero"
+                );
+            }
+
+            // Check that the call path for the segment matches that of the
+            // caller.
+            for (uint256 j = 0; j < _callPath.length - 1; j++) {
+                require(
```
Comment (ricardolyn) [inline functioncall/gpact/src/main/solidity/CrosschainControl.sol:751]: same here

File: functioncall/gpact/src/main/solidity/CrosschainControl.sol (hunk 2 above)
Comment (ricardolyn) [reply functioncall/gpact/src/main/solidity/CrosschainControl.sol:730]: do we have tests that handle those requires? not the happy path

**Review by ricardolyn — COMMENTED** [review]
could we move the tests from being in the example to be next to the code? we should move examples to a separate folder structure and unit tests should close to the app code

**Comment by ricardolyn** [comment]
it seems that we aren't running any tests during PR. could we add that in GitHub actions?

File: application/atomic-appcontracts/erc20/src/main/java/net/consensys/gpact/appcontracts/atomic/erc20/soliditywrappers/GpactERC20Bridge.java (hunk 1 above)
Comment (drinkcoffee) [reply application/atomic-appcontracts/erc20/src/main/java/net/consensys/gpact/appcontracts/atomic/erc20/soliditywrappers/GpactERC20Bridge.java:43]: Hmmm.... I did a build immediately before the push.... I will do another build. I am pretty sure the meta-data at the end of the contract binary includes a time stamp. I wonder if this is what the issue is? I will investigate.

File: functioncall/gpact/src/main/solidity/CrosschainControl.sol (hunk 2 above)
Comment (drinkcoffee) [reply functioncall/gpact/src/main/solidity/CrosschainControl.sol:730]: @ermyas , good point. This check should happen outside the loop.

Comment (drinkcoffee) [reply functioncall/gpact/src/main/solidity/CrosschainControl.sol:730]: @ricardolyn, we need to do a comprehensive set of tests for the code base. This is an area where these is only happy path testing.

File: functioncall/gpact/src/main/solidity/CrosschainControl.sol (hunk 3 above)
Comment (drinkcoffee) [reply functioncall/gpact/src/main/solidity/CrosschainControl.sol:743]: segCallPath is updated each time through the loop

File: functioncall/gpact/src/main/solidity/CrosschainControl.sol (hunk 4 above)
Comment (drinkcoffee) [reply functioncall/gpact/src/main/solidity/CrosschainControl.sol:751]: segCallPath is updated each time through the loop. Hence, this needs to execute.

**Comment by drinkcoffee** [comment]
> it seems that we aren't running any tests during PR. could we add that in GitHub actions?

If you click on "All Checks" above, you will see the Circle CI jobs that have run. As you will see all tests are executed.

**Comment by drinkcoffee** [comment]
> could we move the tests from being in the example to be next to the code? we should move examples to a separate folder structure and unit tests should close to the app code

The example applications are in a separate directory. There are currently no unit tests for the crosschain control. We need to create a comprehensive set of tests to test this part of the code base.

File: functioncall/gpact/src/main/solidity/CrosschainControl.sol (hunk 2 above)
Comment (ricardolyn) [reply functioncall/gpact/src/main/solidity/CrosschainControl.sol]: but why if we are adding this new checks, we cannot add the tests at the same time? doing it after will create a higher probability of getting lost and missing edge cases.

File: functioncall/gpact/src/main/solidity/CrosschainControl.sol (hunk 4 above)
Comment (ricardolyn) [reply functioncall/gpact/src/main/solidity/CrosschainControl.sol:771]: the question was: where are the new tests to test this edge case?
